instrumentation/jvm-metrics/src/main/java/com/splunk/opentelemetry/instrumentation/jvmmetrics/AllocatedMemoryMetrics.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright Splunk Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.splunk.opentelemetry.instrumentation.jvmmetrics;
+
+import com.sun.management.ThreadMXBean;
+import io.micrometer.core.instrument.Gauge;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.binder.BaseUnits;
+import io.micrometer.core.instrument.binder.MeterBinder;
+import java.lang.management.ManagementFactory;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+public class AllocatedMemoryMetrics implements MeterBinder {
+  private final boolean hasComSunThreadMXBean = hasComSunThreadMXBean();
+
+  @Override
+  public void bindTo(MeterRegistry registry) {
+    if (!hasComSunThreadMXBean) {
+      return;
+    }
+
+    if (!isThreadAllocatedMemoryEnabled()) {
+      return;
+    }
+
+    AllocationTracker allocationTracker = new AllocationTracker();
+    Gauge.builder("jvm.experimental.memory.allocated", allocationTracker::getAllocated)
+        .description("Approximate sum of heap allocations")
+        .baseUnit(BaseUnits.BYTES)
+        .register(registry);
+  }
+
+  private static boolean hasComSunThreadMXBean() {
+    try {
+      Class.forName(
+          "com.sun.management.ThreadMXBean", false, AllocatedMemoryMetrics.class.getClassLoader());
+      return true;
+    } catch (ClassNotFoundException e) {
+      return false;
+    }
+  }
+
+  private static boolean isThreadAllocatedMemoryEnabled() {
+    ThreadMXBean threadBean = (ThreadMXBean) ManagementFactory.getThreadMXBean();
+
+    try {
+      threadBean.getAllThreadIds();
+      return threadBean.isThreadAllocatedMemorySupported()
+          && threadBean.isThreadAllocatedMemoryEnabled();
+    } catch (Error error) {
+      // An error will be thrown for unsupported operations
+      // e.g. SubstrateVM does not support getAllThreadIds
+      return false;
+    }
+  }
+
+  private static class AllocationTracker {
+    private final ThreadMXBean threadBean = (ThreadMXBean) ManagementFactory.getThreadMXBean();
+    // map from thread id -> last measured allocated bytes
+    private final Map<Long, Long> allocatedByThread = new HashMap<>();
+    // allocated bytes by threads that have terminated
+    private long allocatedByDeadThreads = 0;
+
+    long getAllocated() {
```

Comment:
Is there a guarantee that this method will only ever be invoked by one thread? The `HashMap` is not thread-safe.

Reply:
At least with signalfx metrics this is always invoked periodically from the same thread.